ouroboros-consensus/docs/future-ledger-hd.md

```diff
@@ -0,0 +1,258 @@
+`Shelley.Rules.Rupd` and `Shelley.Rules.Tick`) is that on each ticking we pulse
+a chunk of the rewards update so that when we reach the epoch boundary we want
+to have pulsed through the whole `ssStakeGo` snapshot that was used when
+creating the pulser.
+
+In the end, the reward computation has to produce a `RewardUpdate`:
+```haskell
+data RewardUpdate c = RewardUpdate
+  { deltaT :: !DeltaCoin
+  , deltaR :: !DeltaCoin
+  , rs :: !(Map (Credential 'Staking c) (Set (Reward c)))
+  , deltaF :: !DeltaCoin
+  , nonMyopic :: !(NonMyopic c)
+  }
+```
+
+where we will use the `delta*` fields to update the treasury, reserves and fee
+pot on the `NEWEPOCH` rule. We also use the `rs` field when calling
+`updateRewards` which happens also on `NEWEPOCH`.
+
+
+If Consensus can compute the `RewardUpdate` (possibly on a separate thread that
+traverses the map at its own pace?) then we can provide the `RewardUpdate` to
+the `NEWEPOCH` rule so that the pots can be updated.
+
+> **_PLAN:_** Consensus will compute the rewards outside of the Ledger and will
+> provide the parts of the `RewardUpdate` to the ledger. In particular, on the
+> epoch boundary, it will provide `deltaT`, `deltaR`, and `deltaF`. Ledger will
+> not compute the `RewardUpdate` thus the pulser becomes dead code.
+
+> Moreover, the ledger will provide a function `f :: Block -> Set
+> (StakeCredential c)` so that Consensus can supply Ledger with only the part of
+> the `UMap` (the unified map) that it requires. In particular, Consensus will
+> have to apply the `rs :: !(Map (Credential 'Staking c) (Set (Reward c)))` field
+> of the reward update to the unified map on the epoch boundary, prior to
+> computing this view for the ledger (to ensure that reward withdrawals are
+> correct within the given block).
+
+## Snapshots and Leader Schedule
+
+The `checkIsLeader` functions for Praos and TPraos makes use of the stake
+distribution by stake pool in the `LedgerView` (see the definitions in
+`ouroboros-consensus-protocol`). If the snapshots (and therefore the
+`ssStakeMarkPoolDistr` field) reside in the Consensus side, we can produce the
+relevant stake distributions when needed and don't involve the ledger. In any
+case this functionality is in between Ledger and Consensus so it makes sense to
+move it out of the ledger.
+
+> **_PLAN:_** Consensus will manage the snapshots to produce stake distribution
+> by pool that can be used by Consensus later to resolve queries about the
+> LeaderSchedule. Ledger will not know about the Snapshots. In particular, the
+> UTxO-HD report includes the concept of Snapshots of tables, which would be
+> used to manage and access these snapshots.
+
+Note that this implies creating a new package or component at the
+Consensus-Ledger boundary whose owner would probably be the Consensus team as
+its responsibilities would be related with computations required for the
+Consensus protocol (leader checks, and similar).
```

Comment:
I'm slightly unclear about what this section is porposing. The stake distribution by pool is not large, so will not be an on-disk table. It will remain as an in-memory structure as it is now, within the ledger state and ledger view. I don't see the need for any consensus/ledger interface changes.

Reply:
I commented on this above.